src/plugins/workflow_designer/src/library/ReadAnnotationsWorker.cpp

```diff
@@ -242,30 +281,34 @@ void ReadAnnotationsTask::run() {
     m[BaseSlots::URL_SLOT().getId()] = url;
     m[BaseSlots::DATASET_SLOT().getId()] = datasetName;
 
-    QList<SharedAnnotationData> dataList;
-
-    for (GObject *go : qAsConst(annsObjList)) {
-        auto annsObj = dynamic_cast<AnnotationTableObject *>(go);
-        CHECK_EXT(annsObj != nullptr, stateInfo.setError("NULL annotations object"), );
-
-        if (!mergeAnnotations || annsObjList.size() == 1) {
-            doc->removeObject(go, DocumentObjectRemovalMode_Release);
-            const SharedDbiDataHandler tableId = context->getDataStorage()->putAnnotationTable(annsObj);
-            delete go;
+    // If "SPLIT" -- transfer (write to the result) tables of annotations from files as is.
+    // If "MERGE" and there is only one annotation table, transfer the table as is.
+    // If "MERGE_FILES", transfer the file tables as is and merge them into ReadAnnotationsWorker::sl_datasetEnded().
+    // Otherwise ("MERGE" with several tables of annotations in one file), merge these tables into one.
+    if (mergeAnnotations != ReadAnnotationsProto::MERGE || annsObjList.size() == 1) {
+        for (GObject *gObject : qAsConst(annsObjList)) {
+            auto annotationTable = dynamic_cast<AnnotationTableObject *>(gObject);
+            CHECK_EXT(annotationTable != nullptr, setError(L10N::nullPointerError("annotationTable")), );
+
+            QScopedPointer<AnnotationTableObject> annotationTableClone(qobject_cast<AnnotationTableObject *>(annotationTable->clone(dbiRef, stateInfo)));
```

Comment:
Why do you want to clone here? Clone() creates another instance in DB and if we have only 1 Annotation Object we can just return it, right?

Reply:
How to transfer a table to another worker? It will be removed with the document.

Reply:
All workers in workflow use the same share database and pass data to each other via QSharedDataPointer<DbiDataHandler>

Here how you create your annotations:
`
hints[DocumentFormat::DBI_REF_HINT] = QVariant::fromValue<U2DbiRef>(context->getDataStorage()->getDbiRef());
QScopedPointer<Document> doc(format->loadDocument(iof, url, hints, stateInfo));
`
Here how you clone it several lines below:
`
go->clone(context->getDataStorage(), stateInfo)))
`

It is the same database already: no need to clone, just re-use the object you loaded.
You should also ensure that the document won't destroy these objects in the database during own destruction: remove the objects from the document.

Reply:
How to delete objects from the document?

Reply:
```bool Document::removeObject``` ? 
If the existing API is not enough you can extend it.